frontend/public/components/utils/category-filter-utils.ts

```diff
@@ -0,0 +1,391 @@
+/* eslint-disable no-undef */
+import * as _ from 'lodash-es';
+import {history} from './router';
+
+const CATEGORY_URL_PARAM = 'category';
+const KEYWORD_URL_PARAM = 'keyword';
+const DETAILS_ITEM_URL_PARAM = 'details-item';
+
+const filterSubcategories = (category: any, item: any) => {
+  if (!_.size(category.subcategories)) {
+    if (!category.values) {
+      return [];
+    }
+
+    let values = _.get(item, category.field);
+    if (!Array.isArray(values)) {
+      values = [values];
+    }
+
+    if (!_.isEmpty(_.intersection(category.values, values))) {
+      return [category];
+    }
+
+    return [];
+  }
+  const matchedSubcategories = [];
+
+  _.forEach(category.subcategories, subCat => {
+    let values = _.get(item, category.field);
+    if (!Array.isArray(values)) {
+      values = [values];
+    }
+    if (!_.isEmpty(_.intersection(subCat.values, values))) {
+      matchedSubcategories.push(subCat);
+      matchedSubcategories.push(...filterSubcategories(subCat, item));
+    }
+  });
+
+  return matchedSubcategories;
+};
+
+// categorize item under sub and main categories
+const addItem = (item: any, category: any, subcategory: any = null) => {
```

Comment:
We should be defining types for item, category, and subcategory if we're pulling this stuff out as utilities. We can benefit from type checking here.